File: iRead/iReader/Main/Main/IRMainViewController.m
//
//  MainViewController.m
//  iReader
//
//  Created by zouzhiyong on 2018/3/12.
//  Copyright © 2018年 zouzhiyong. All rights reserved.
//

#import "IRMainViewController.h"
#import "IRHomeViewController.h"
#import "IRMineViewController.h"

@interface IRMainViewController () <UITabBarControllerDelegate>


@end

@implementation IRMainViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self commonInit];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

#pragma mark - setter/getter

- (void)setSelectedIndex:(NSUInteger)selectedIndex
{
    [super setSelectedIndex:selectedIndex];
    [self updateNavigationItemsForIndex:selectedIndex];
}

#pragma mark - UITabBarControllerDelegate

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController
{
    [self updateNavigationItemsForIndex:[tabBarController.viewControllers indexOfObject:viewController]];
}

#pragma mark - private

- (void)commonInit
{
    self.delegate = self;
    [self setupTabbarItems];
    if (self.selectedIndex != 0) {
        self.selectedIndex = 0;
    }
}

- (void)setupTabbarItems
{
    self.tabBar.tintColor = APP_THEME_COLOR;
    NSArray<NSString *> *tabbarTitles = @[@"首页", @"我的"];
    NSMutableArray *childViewControllers = [NSMutableArray arrayWithCapacity:tabbarTitles.count];
    [tabbarTitles enumerateObjectsUsingBlock:^(NSString * _Nonnull title, NSUInteger idx, BOOL * _Nonnull stop) {
        [childViewControllers addObject:[self childViewControllerWithTabIndex:idx]];
    }];
    self.viewControllers = childViewControllers;
    
    [self.tabBar.items enumerateObjectsUsingBlock:^(UITabBarItem * _Nonnull item, NSUInteger idx, BOOL * _Nonnull stop) {
        
        UIImage *normalImg = nil;
        UIImage *selectImg = nil;
        if (idx == IRMainTabIndexHome) {
            normalImg = [UIImage imageNamed:@"tabbar_home_n"];
            selectImg = [[UIImage imageNamed:@"tabbar_home_s"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
        } else if (idx == IRMainTabIndexMine) {
            normalImg = [UIImage imageNamed:@"tabbar_mine_n"];
            selectImg = [[UIImage imageNamed:@"tabbar_mine_s"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
        }
        item.title = tabbarTitles[idx];
        item.image = normalImg;
        item.selectedImage = selectImg;
    }];
}

- (UIViewController *)childViewControllerWithTabIndex:(IRMainTabIndex)tabIndex
{
    UIViewController *vc = nil;
    
    if (tabIndex == IRMainTabIndexHome) {
        vc = [[IRHomeViewController alloc] init];
    } else if (tabIndex == IRMainTabIndexMine) {
        vc = [[IRMineViewController alloc] init];
    } else {
        vc = [[UIViewController alloc] init];
        NSAssert(NO, @"unsupport type");
    }
    return vc;
}

- (void)updateNavigationItemsForIndex:(IRMainTabIndex)index
{
    self.navigationItem.rightBarButtonItem  = nil;
    self.navigationItem.rightBarButtonItems = nil;
    self.navigationItem.leftBarButtonItem   = nil;
    
    switch (index) {
        case IRMainTabIndexHome: {
            self.navigationItem.title = self.selectedViewController.navigationItem.title;
            self.navigationItem.rightBarButtonItem = self.selectedViewController.navigationItem.rightBarButtonItem;
            break;
        }
            
        case IRMainTabIndexMine: {
            self.navigationItem.title = self.selectedViewController.navigationItem.title;
            break;
        }
    }
}

@end

File: iRead/iRead/Main/Main/IRMainViewController.swift
//
//  IRMainViewController.swift
//  iRead
//
//  Created by zzyong on 2020/9/24.
//  Copyright © 2020 zzyong. All rights reserved.
//

import UIKit
import IRCommonLib

enum IRTabBarIndex: Int {
    case home      = 0
    case bookshelf = 1
    case mine      = 2
}

enum IRTabBarName: String {
    case home      = "首页"
    case bookshelf = "书架"
    case explore   = "探索"
}

class IRMainViewController: UITabBarController, UITabBarControllerDelegate {

    var initOnceAfterViewDidAppear = false
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.commonInit()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.initAfterViewDidAppear()
    }
    
    override var selectedIndex: Int {
        willSet {
            self.updateNavigationItems(withIndex: newValue)
        }
    }
    
    // MARK: - Private
    
    func initAfterViewDidAppear() {
        if initOnceAfterViewDidAppear {
            return
        }
        initOnceAfterViewDidAppear = true
        self.updateReadPageSzie()
        IRReaderConfig.initReaderConfig()
    }
    
    func updateReadPageSzie() {
        
        var safeInsets = UIEdgeInsets.zero
        if #available(iOS 11.0, *) {
            safeInsets = self.view.safeAreaInsets
        }
        
        if safeInsets.bottom == 0 || safeInsets.top == 0 {
            safeInsets = UIEdgeInsets.init(top: 30, left: 0, bottom: 30, right: 0)
        }
        
        let width = self.view.width - IRReaderConfig.horizontalSpacing * 2
        let height = self.view.height - safeInsets.top - safeInsets.bottom - IRReaderConfig.pageIndexSpacing
        let maxSize: CGFloat = 1000
        IRReaderConfig.pageSzie = CGSize.init(width: min(maxSize, width), height: min(maxSize, height))
    }
    
    func commonInit() {
        self.delegate = self
        self.setupTabbarItems()
    }
    
    func setupTabbarItems() {
        self.tabBar.tintColor = IRAppThemeColor
        
        let tabbarTitles = [IRTabBarName.home.rawValue, IRTabBarName.bookshelf.rawValue, IRTabBarName.explore.rawValue]
        var childViewControllers = [UIViewController]()
        
        for (index, _) in tabbarTitles.enumerated() {
            childViewControllers.append(self.childViewController(withTabIndex: index))
        }
        self.viewControllers = childViewControllers
        
        for (index, item) in self.tabBar.items!.enumerated() {
            var normalName: String?, selectName: String?
            switch index {
            case IRTabBarIndex.home.rawValue:
                normalName = "tabbar_home_n"; selectName = "tabbar_home_s"
            case IRTabBarIndex.bookshelf.rawValue:
                normalName = "tabbar_bookshelf_n"; selectName = "tabbar_bookshelf_s"
            case IRTabBarIndex.mine.rawValue:
                normalName = "tabbar_explore_n"; selectName = "tabbar_explore_s"
            default:
                IRDebugLog("TabIndex: (\(index)) undefine")
            }
            item.title = tabbarTitles[index]
            if let normalName = normalName, let selectName = selectName {
                item.image = UIImage.init(named: normalName)?.withRenderingMode(.alwaysOriginal)
                item.selectedImage = UIImage.init(named: selectName)?.withRenderingMode(.alwaysOriginal)
            }
        }
    }
    
    func childViewController(withTabIndex index: Int) -> UIViewController {
        var vc: UIViewController
        switch index {
        case IRTabBarIndex.home.rawValue:
            vc = IRHomeViewController()
        case IRTabBarIndex.bookshelf.rawValue:
            vc = IRBookshelfViewController()
        case IRTabBarIndex.mine.rawValue:
            vc = IRExploreViewController()
        default:
            vc = UIViewController()
        }
        return vc
    }
    
    func updateNavigationItems(withIndex index: Int) {
        self.navigationItem.rightBarButtonItems = nil
        self.navigationItem.leftBarButtonItems = nil
        
        self.navigationItem.title = self.selectedViewController?.navigationItem.title
    }
    
    // MARK: - UITabBarControllerDelegate
    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        self.updateNavigationItems(withIndex: tabBarController.viewControllers?.firstIndex(of:viewController) ?? IRTabBarIndex.home.rawValue)
    }
}
